pydarn/exceptions/borealis_exceptions.py

```diff
@@ -0,0 +1,156 @@
+            " {fields}".format(num=self.record_name,
+                               fields=self.fields)
+        Exception.__init__(self, self.message)
+
+
+class BorealisExtraFieldError(Exception):
+    """
+    Raised when a extra field is in the data that is not in the Borealis file
+    type fields
+
+    Parameter
+    ---------
+    record_name : int
+        record name of the problem record
+    fields : set
+        set of fields that are missing from the record
+    """
+
+    def __init__(self, record_name: int, fields: set):
+        self.record_name = record_name
+        self.fields = fields
+        self.message = "The following fields in record {num} are not allowed:"\
+            " {fields}".format(num=self.record_name,
+                               fields=self.fields)
+        Exception.__init__(self, self.message)
+
+
+class BorealisDataFormatTypeError(Exception):
+    """
+    Raised when a data format type does not match the Borealis file type
+    data types
+
+    Parameter
+    ---------
+    incorrect_type : dict
+        set of the fields that have incorrect types
+    record_name : int
+        record name of the problem record
+    """
+
+    def __init__(self, incorrect_types: set, record_name: int):
+        self.incorrect_params = incorrect_types
+        self.record_name = record_name
+        self.message = "In record {num}, the following parameters"\
+            " need to be the data type:"\
+            " {incorrect}".format(num=self.record_name,
+                                  incorrect=self.incorrect_params)
+        Exception.__init__(self, self.message)
+
+
+class BorealisConversionTypesError(Exception):
+    """
+    SuperDARN Borealis filetype cannot be converted to desired type.
+
+    Parameter
+    ---------
+    filename : str
+        name of the file attempting to convert
+    origin_filetype: str
+        Borealis origin filetype, ex. bfiq.
+    dmap_filetype : str
+        desired SuperDARN legacy dmap type, ex. iqdat.
+    """
+
+    def __init__(self, filename: str, origin_filetype: str, dmap_filetype: str):
+        self.filename = filename
+        self.origin_filetype = origin_filetype
+        self.dmap_filetype = dmap_filetype
+        self.message = "{filename} cannot be converted from its origin_filetype "\
+                       "{origin_filetype} to legacy dmap filetype {dmap_filetype} because the "\
+                       "types are not compatible or currently not supported.".format(filename=self.filename,
+                                                                                     origin_filetype=self.origin_filetype, dmap_filetype=self.dmap_filetype)
+        Exception.__init__(self, self.message)
+
+
+class BorealisConvert2IqdatError(Exception):
+    """
+    Raised when the file cannot be converted to legacy DARN dmap iqdat format.
+
+    Parameter
+    ---------
+    error_str : str
+        explanation for why the file cannot be converted to legacy DARN iqdat.
```

Comment:
I thought iqdat is a filetype of the DARN class, dmap could have any filetype not specified. Which do you prefer?

Reply:
The lingo of SuperDARN. I think SuperDARN RAWACF or IQDAT would be better. I think DARN is vague without knowing my DARN class which is under superdarn.py. 

DMap is a file format.

Reply:
I think I prefer the reference to the class though instead of being general. The SuperDARN rawacf standard format may change in the future for example; referencing the DARN class gives information of exactly what conversion failed. Could I include a See Also : DARNWrite to make it more clear?

Reply:
True but wouldn't it still be called a SuperDARN Rawacf and DARN class would just be updated to that new standard format. 

Anyways, I am still cool with see Also or something referencing Darn classes as that is a bit more explicit.

Reply:
maybe, but I think inevitably there will be a quite lengthy transition period (decades I'd guess) during which there may exist two standards. Many radars will continue to operate ROS, so we will need to support this DARN dmap format continuing into the future, even if a new standard also becomes available. Since we will need to be converting between the two, I think the DARN class will be necessary, since it's not reasonable to expect groups to change the ROS software writing the rawacf dmap format at the radar overnight.